### PR TITLE
Semantics and test for ENTRY statements

### PR DESCRIPTION
**Pull request by klausler**
Checks the constraints and the feasible "shalls" in 15.6.2.6 for `ENTRY` statements, their dummy arguments, and their results.

Symbols for the entry points are created in the parent scope of the containing subprogram, and point to the scope of the containing subprogram.  Symbols for `ENTRY` results, when functions, are within the scope of the containing subprogram.

This PR checks semantics only; it doesn't do anything to represent sharing of dummy arguments, storage association of function results, &c.

### CODE REVIEW

**Comment by klausler** [comment]
Testing against f90_correct shows that there's still some bugs to fix in this code, so updates are coming.  I'll comment again when all tests are passing.  Also, I need to figure out how to emit `ENTRY` points in module files.

**Comment by klausler** [comment]
Ok, the tests are greatly extended to cover additional cases, and they pass; so do all of the cases in the PGI Fortran suite.  And I've figured out how I think `ENTRY` should interact with module files.

File: include/flang/Semantics/symbol.h [hunk 1]
```diff
@@ -82,8 +85,10 @@ class SubprogramDetails {
   MaybeExpr bindName_;
   std::vector<Symbol *> dummyArgs_;  // nullptr -> alternate return indicator
   Symbol *result_{nullptr};
+  Scope *entryScope_{nullptr};  // if ENTRY, points to subprogram's scope
```
Comment (jeanPerier) [inline include/flang/Semantics/symbol.h]: I am a bit lost with the purpose of entry_scope() vs symbol->scope().
Is this to keep the property that `symbol->scope()->symbol() = identity` or does it have other purposes ?

Comment (klausler) [reply include/flang/Semantics/symbol.h]: `Symbol::scope()` is used by subprograms, derived types, modules, &c. to link symbols to the scopes that they own.  `ENTRY` point symbols don't have their own scopes, and they are not in the scope of the subprogram in which they appear.  For example,
```
subroutine subr
real x
entry ent
end
```
there's the global scope, which contains the symbol subr` that owns the scope that contains `x`.  The symbol `ent` is also in the global scope -- to its callers, it's no different from any other subroutine -- but it doesn't own a scope.  Instead, it links to the scope of `subr` via its `entryScope_` data member.

Comment (jeanPerier) [reply include/flang/Semantics/symbol.h]: Makes sens, thanks for explaining.

**Pull request merged**